app/services/payment_requests/payments/adyen_service.rb

```diff
@@ -76,6 +76,29 @@ def generate_payment_url
         result.service_failure!(code: e.code, message: e.msg)
       end
 
+      def update_payment_status(provider_payment_id:, status:, metadata: {})
+        payment = if metadata[:payment_type] == "one-time"
+          create_payment(provider_payment_id:, metadata:)
+        else
+          Payment.find_by(provider_payment_id:)
+        end
+        return result.not_found_failure!(resource: "adyen_payment") unless payment
+
+        result.payment = payment
+        result.payable = payment.payable
+        return result if payment.payable.payment_succeeded?
+
+        payment.update!(status:)
+
+        payable_payment_status = payable_payment_status(status)
+        update_payable_payment_status(payment_status: payable_payment_status)
+        update_invoices_payment_status(payment_status: payable_payment_status)
```

Comment:
We could probably do this inside a transaction as well (`ActiveRecord::Base.transaction`)